stackit/internal/services/iaas/machinetype/datasource.go

```diff
@@ -0,0 +1,237 @@
+				Description: `Expr-lang filter for filtering machine types.
+
+Examples:
+- vcpus == 2
+- ram >= 2048
+- extraSpecs.cpu == "intel-icelake-generic"
+- extraSpecs.cpu == "intel-icelake-generic" && vcpus == 2
+
+See https://expr-lang.org/docs/language-definition for syntax.`,
+				Required: true,
+			},
+			"description": schema.StringAttribute{
+				Description: "Machine type description.",
+				Computed:    true,
+			},
+			"disk": schema.Int64Attribute{
+				Description: "Disk size in GB.",
+				Computed:    true,
+			},
+			"extra_specs": schema.MapAttribute{
+				Description: "Extra specs (e.g., CPU type, overcommit ratio).",
+				ElementType: types.StringType,
+				Computed:    true,
+			},
+			"name": schema.StringAttribute{
+				Description: "Name of the machine type (e.g. 's1.2').",
+				Computed:    true,
+			},
+			"ram": schema.Int64Attribute{
+				Description: "RAM size in MB.",
+				Computed:    true,
+			},
+			"vcpus": schema.Int64Attribute{
+				Description: "Number of vCPUs.",
+				Computed:    true,
+			},
+		},
+	}
+}
+
+func (m *machineTypeDataSource) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) { // nolint:gocritic // function signature required by Terraform
+	var model DataSourceModel
+	resp.Diagnostics.Append(req.Config.Get(ctx, &model)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	projectId := model.ProjectId.ValueString()
+	sortAscending := model.SortAscending.ValueBool()
+
+	ctx = tflog.SetField(ctx, "project_id", projectId)
+	ctx = tflog.SetField(ctx, "filter_is_null", model.Filter.IsNull())
+	ctx = tflog.SetField(ctx, "filter_is_unknown", model.Filter.IsUnknown())
+
+	listMachineTypeReq := m.client.ListMachineTypes(ctx, projectId)
+
+	if !model.Filter.IsNull() && !model.Filter.IsUnknown() && strings.TrimSpace(model.Filter.ValueString()) != "" {
+		listMachineTypeReq = listMachineTypeReq.Filter(strings.TrimSpace(model.Filter.ValueString()))
+	}
+
+	apiResp, err := listMachineTypeReq.Execute()
+	if err != nil {
+		utils.LogError(ctx, &resp.Diagnostics, err, "Failed to read machine types",
+			fmt.Sprintf("Unable to retrieve machine types for project %q %s.", projectId, err),
+			map[int]string{
+				http.StatusForbidden: fmt.Sprintf("Access denied to project %q.", projectId),
+			},
+		)
+		resp.State.RemoveResource(ctx)
+		return
+	}
+
+	if apiResp.Items == nil || len(*apiResp.Items) == 0 {
+		core.LogAndAddWarning(ctx, &resp.Diagnostics, "No machine types found", "No matching machine types.")
+		return
+	}
+
+	// Convert items to []*iaas.MachineType
+	machineTypes := make([]*iaas.MachineType, len(*apiResp.Items))
+	for i := range *apiResp.Items {
+		machineTypes[i] = &(*apiResp.Items)[i]
+	}
+
+	sorted, err := sortMachineTypeByName(machineTypes, sortAscending)
+	if err != nil {
+		core.LogAndAddWarning(ctx, &resp.Diagnostics, "Unable to sort", err.Error())
+		return
+	}
+
+	first := sorted[0]
+	if err := mapDataSourceFields(ctx, first, &model); err != nil {
```

Comment:
nitpick: can be simplified to one line
```suggestion
	if err := mapDataSourceFields(ctx, sorted[0], &model); err != nil {
```